securebanking-openbanking-uk-gateway-conversion-filter/src/main/java/com/forgerock/securebanking/uk/gateway/conversion/filter/IntentConverterFilter.java

```diff
@@ -0,0 +1,219 @@
+import org.forgerock.http.header.ContentTypeHeader;
+import org.forgerock.http.protocol.*;
+import org.forgerock.json.JsonValue;
+import org.forgerock.json.JsonValueException;
+import org.forgerock.openig.heap.GenericHeaplet;
+import org.forgerock.openig.util.MessageType;
+import org.forgerock.services.context.Context;
+import org.forgerock.util.Function;
+import org.forgerock.util.Strings;
+import org.forgerock.util.promise.NeverThrowsException;
+import org.forgerock.util.promise.Promise;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+
+import static com.forgerock.securebanking.uk.gateway.utils.IntentTypeUtils.getIntentType;
+import static org.forgerock.http.protocol.Entity.APPLICATION_JSON_CHARSET_UTF_8;
+import static org.forgerock.json.JsonValueFunctions.enumConstant;
+import static org.forgerock.util.promise.Promises.newResultPromise;
+
+/**
+ * Filter to convert IDM json intent objects to OB data model objects.
+ *
+ * This filter must have received {@link  MessageType} as required to get the entity  to instance the converter and know the {@link MessageImpl} to be updated<br/>
+ * Failures from the `Converter instance` is either a {@link RuntimeException} that will catch to build a {@link ResponseException}.
+ *
+ * Configuration options:
+ *
+ * <pre>
+ * {@code {
+ *      "name": "IntentConverterFilter-name"
+ *      "type": "IntentConverterFilter",
+ *      "config": {
+ *         "messageType"  MessageType   [REQUIRED - The type of message for which to convert the entity, Must be either "REQUEST" or "RESPONSE"]
+ *      }
+ *  }
+ *  }
+ * </pre>
+ * <p>Examples</p>
+ * <pre>
+ * {@code {
+ *      "name": "IntentConverterFilter-name"
+ *      "type": "IntentConverterFilter",
+ *      "config": {
+ *         "messageType: "request"
+ *      }
+ *  }
+ *  }
+ * </pre>
+ * <pre>
+ *  {@code {
+ *       "name": "IntentConverterFilter-name"
+ *       "type": "IntentConverterFilter",
+ *       "config": {
+ *          "messageType: "response"
+ *       }
+ *   }
+ *   }
+ *  </pre>
+ */
+public class IntentConverterFilter implements Filter {
+
+    private static final Logger logger = LoggerFactory.getLogger(IntentConverterFilter.class);
+
+    private final MessageType messageType;
+    private static final ObjectMapper MAPPER = GenericConverterMapper.getMapper();
+
+    /**
+     * Constructor
+     * @param messageType
+     */
+    public IntentConverterFilter(final MessageType messageType) {
+        this.messageType = messageType;
+    }
+
+    @Override
+    public Promise<Response, NeverThrowsException> filter(Context context, Request request, Handler next) {
+        if (logger.isInfoEnabled()) {
+            printFilterInfo();
+        }
+
+        if (messageType.equals(MessageType.REQUEST)) {
+            return process(request, request.getUri().asURI(), context, next);
+        }
+
+        return next.handle(context, request)
+                .thenAsync(response -> {
```

Comment:
then can be used rather than thenAsync, as there is no async call happening in the callback. The callback applies a transformation to data which is available.

Also, you are ignoring the return value from the process method, therefore if we were to hit the catch block in that method the error response that gets created is lost.

I think it would be cleaner to extract a method which transforms the message.entity i.e. lines 130-142, then leave it up to the caller to plug the logic into the request or response flow. As the process method has become quite complex, having to deal with request and response specific concerns.

Reply:
Improvements pushed.

Reply:
That's excellent, I think that code is much cleaner, simpler and easier to understand now.

I've added a couple of minor comments to do some cleanup, once that is done then I think it's ready to merge